bufferevent_mbedtls.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2009-2012 Niels Provos and Nick Mathewson
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <mbedtls/ssl.h>
+#include <mbedtls/net_sockets.h>
+#include <mbedtls/error.h>
+
+#include "event2/util.h"
+#include "util-internal.h"
+#include "event2/buffer.h"
+#include "event2/bufferevent.h"
+#include "event2/bufferevent_struct.h"
+
+#define SSL mbedtls_ssl_context
+#define SSL_ERROR_WANT_READ MBEDTLS_ERR_SSL_WANT_READ
+#define SSL_ERROR_WANT_WRITE MBEDTLS_ERR_SSL_WANT_WRITE
+#define SSL_free mbedtls_ssl_free
+#define SSL_renegotiate mbedtls_ssl_renegotiate
+#define SSL_pending mbedtls_ssl_get_bytes_avail
+#define SSL_write mbedtls_ssl_write
+#define SSL_read mbedtls_ssl_read
+#define SSL_do_handshake mbedtls_ssl_handshake
+#define SSL_is_ok(x) (x == 0)
+#define SSL_get_error(ssl, x) (x)
```

Comment:
I don't like the idea of using `SSL_*` here (since it overlaps with openssl, I understand that by using this you don't need to provide same thing for openssl, but looks error prone), let's name it somehow different, i.e. `LE_SSL_*` (or maybe you have better name for it)

Reply:
you mean all macro replacement？eg. SSL_do_handshake.

or just macro function? eg. SSL_is_ok

Reply:
I have better suggestion - https://github.com/libevent/libevent/pull/1028#discussion_r452044652